module/zfs/dsl_scan.c

```diff
@@ -137,6 +137,16 @@ extern int zfs_vdev_async_write_active_min_dirty_percent;
  */
 int zfs_scan_strict_mem_lim = B_FALSE;
 
+/*
+ * Maximum number of parallelly executing I/Os per top-level vdev.
+ * Tune with care. Very high settings (hundreds) are known to trigger
+ * some firmware bugs and resets on certain SSDs.
+ */
+int zfs_top_maxinflight = 32;		/* maximum I/Os per top-level */
```

Comment:
Um, I don't see this one being actually used anywhere...

Reply:
Looks like I missed the actual use of it:
```
/* ARGSUSED */
static void
dsl_scan_prefetch_thread(void *arg)
{
	dsl_scan_t *scn = arg;
	spa_t *spa = scn->scn_dp->dp_spa;
	vdev_t *rvd = spa->spa_root_vdev;
	uint64_t maxinflight = rvd->vdev_children * zfs_top_maxinflight;
	scan_prefetch_issue_ctx_t *spic;

	/* loop until we are told to stop */
	while (!scn->scn_prefetch_stop) {
		arc_flags_t flags = ARC_FLAG_NOWAIT |
                    ARC_FLAG_PRESCIENT_PREFETCH | ARC_FLAG_PREFETCH;
		int zio_flags = ZIO_FLAG_CANFAIL | ZIO_FLAG_SCAN_THREAD;
		
		mutex_enter(&spa->spa_scrub_lock);

		/*
		 * Wait until we have an IO to issue and are not above our
		 * maximum in flight limit.
		 */
		while (!scn->scn_prefetch_stop &&
		    (avl_numnodes(&scn->scn_prefetch_queue) == 0 ||
		    spa->spa_scrub_inflight >= scn->scn_maxinflight_bytes)) {
			cv_wait(&spa->spa_scrub_io_cv, &spa->spa_scrub_lock);
		}

```

Reply:
I don't know if these sorts of arbitrary limits make sense. They don't tend to age well. However, I'm not sure how easy it would be to implement some sort of PID controller for it.

Reply:
PID controllers need careful tuning, we'd need to come up with some zfs-tuned or something, I guess...

Reply:
zfs_top_maxinflight did not work well in practice, mostly because it was multiplied by the number of children, so hundreds was not uncommon.  In any case, this throttling mechanism proved to be nearly impossible to optimize in the field.

The current ZIO scheduler approach seems to work better, though the defaults are questionable and biased against scan workloads.